### DISCUSSION TITLE
`mutate` with `optimisticData` clears `data` on revalidation

### DISCUSSION THREAD

**Discussion by cbejensen** [vercel/swr#2603] [Q&A] [answered]
# Bug report

When using `mutate` with `optimisticData` and `revalidate: true`, the `data` is cleared and `isLoading` is set to `true` during revalidation.

## Expected Behavior

I would expect `data` to be preserved while revalidating. Also, I would expect `isValidating` to be `true`, but `isLoading` to be `false`.

## Code Example

https://codesandbox.io/s/swr-basic-forked-6srjsk?file=/src/App.js

**Comment by koba04** [comment, marked as answer]
@cbejensen 
In your `api.js`, `addTodo` returns nothing, and `populateCache` is `true` (the default value of `mutate`), which means that SWR updates its cache data with `undefined`, so `useSWR` returns `undefined` when the next revalidation.

```js
export async function addTodo(todo) {
  await delay();
  todos = [...todos, todo];
}
```

https://swr.vercel.app/docs/mutation#parameters

To fix that, you can disable the `populateCache` option at `mutate` or return the updated data from `addTodo`.

**Comment by cbejensen** [comment]
Ah, just my misunderstanding then - thank you!

**Reply by koba04** [reply]
Thank you for your feedback! 👍
